File: leetcode_problems/cpp/lc1046_last_stone_weight.cpp
#include <vector>
#include <queue>
using std::priority_queue;
using std::vector;

class Solution
{
public:
    int lastStoneWeight(vector<int> &stones)
    {
        priority_queue<int> heap;
        for (int st : stones)
        {
            heap.push(st);
        }
        while (heap.size() > 1)
        {
            int f = heap.top();
            heap.pop();
            int s = heap.top();
            heap.pop();
            if (s != f)
            {
                heap.push(f - s);
            }
        }
        return heap.size() > 0 ? heap.top() : 0;
    }
};

File: leetcode_problems/go/lc1046_last_stone_weight.go
package main

import "container/heap"

type IntHeap []int

func (h IntHeap) Len() int           { return len(h) }
func (h IntHeap) Less(i, j int) bool { return h[i] > h[j] } // max heap condition
func (h IntHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }

func (h *IntHeap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *IntHeap) Pop() interface{} {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

func lastStoneWeight(stones []int) int {
	h := &IntHeap{}
	for _, stone := range stones {
		heap.Push(h, stone)
	}
	heap.Init(h)
	for h.Len() > 1 {
		s1 := heap.Pop(h).(int)
		s2 := heap.Pop(h).(int)
		spread := abs(s1 - s2)
		if spread > 0 {
			heap.Push(h, spread)
		}
	}
	if h.Len() == 0 {
		return 0
	}
	return heap.Pop(h).(int)
}
